Unify build_native via eng/native/build-commons

am11 opened the pull request:
In order to add new platform, architecture, compiler or its newer
version, there are currently multiple places to update, which can be
deduplicated.

This patch unifies:

* directories creation
* prerequisites checking
* `version.c` generation
* native build invocation
* common argument parsing
* building help menu

for various build scripts under coreclr, installer and libraries.

The common entry-point script is now `eng/native/build-commons.sh` and
rest of the scripts under `eng/native` are implementation detail.

Also extracted CMake platform detection in
`eng/native/configureplatform.cmake`, to share with coreclr and
`installer/corehost`.

am11 commented:
@janvorli, arm pri0 test failure looks real going by [helix log](https://helix.dot.net/api/2019-06-17/jobs/5016ddc9-621c-41c7-a589-34adade1a8d5/workitems/Interop/console). Do you have some clues what might be the cause of this assertion:

```sh
Assert failure(PID 133 [0x00000085], Thread: 133 [0x0085]): !"Recursion in CLRException::GetThrowable"
    File: /__w/4/s/src/coreclr/src/vm/clrex.cpp Line: 148
    Image: /root/helix/work/correlation/corerun
```

Alternatively, to investigate it locally using lldb etc., could you tell me the command to run pri0 tests? I have arm build ready via rootfs, I will try to run it on arm device.

janvorli commented:
This error usually indicates mismatch between libcoreclr.so and System.Private.CoreLib.dll. E.g. when these files are built from different source tree states or one of them is of debug /checked and the other release build. 

As for debugging on arm device, what I usually do is to cross-build the tests using the build-test.sh (pass it the same options as to build.sh). Then I clone the same state of the repo on arm, ideally to the same path as on the build device and copy over the artifacts folder from the build machine to the arm one.
Then I run all the tests using the exact command line the build-test prints at the end or run just a single test. For debugging, the easiest way is to run a single test as follows (assuming you are using debug build, which I would highly recommend)
1. cd into the binary folder of the test (where the .sh file for the test resides). It is somewhere under the artifacts/tests/coreclr/Linux.arm.Debug.
3. run gdb --args /full/path/to/corerun/in/tests/core_root your_test_assembly.dll (the corerun is located in artifacts/tests/coreclr/Linux.arm.Debug/Tests/Core_Root)

The .sh file for each test can also be used for starting the test under a debugger (-debug option), but it has a problem with passing arguments to gdb / lldb currently, it doesn't put `--` after lldb and `--args` after gdb that is necessary for correctly passing the arguments to the corerun. So I don't use it when I need to debug stuff.

am11 commented:
Thanks! I could not spot anything obvious in this PR which suggest the configuration mismatch. So I am investigating on arm device.

The error was spotted with Checked build, so I went ahead with that configuration. Here is a stack trace on executing a test on arm device (same path as my az vm running Ubuntu where i cross compiled for arm):

<details>
<summary>click to expand</summary>

```sh
pi@raspberrypi:/datadrive/projects/runtime $ bash artifacts/tests/coreclr/Linux.arm.Checked/JIT/CodeGenBringUpTests/ArrayExc_d/ArrayExc_d.sh -coreroot=$(pwd)/artifacts/bin/coreclr/Linux.arm.Checked
BEGIN EXECUTION
/datadrive/projects/runtime/artifacts/bin/coreclr/Linux.arm.Checked/corerun ArrayExc_d.dll ''
Unhandled exception. Process terminated. Infinite recursion during resource lookup within System.Private.CoreLib.  This may be a bug in System.Private.CoreLib, or potentially in certain extensibility points such as assembly resolve events or CultureInfo names.  Resource name: Word_At
   at System.Environment.FailFast(System.String)
   at System.SR.InternalGetResourceString(System.String)
   at System.SR.GetResourceString(System.String, System.String)
   at System.Diagnostics.StackTrace.ToString(TraceFormat, System.Text.StringBuilder)
   at System.Diagnostics.StackTrace.ToString(TraceFormat)
   at System.Diagnostics.DebugProvider.Fail(System.String, System.String)
   at System.Globalization.CultureInfo.get_InvariantCulture()
   at System.Resources.ManifestBasedResourceGroveler.GetNeutralResourcesLanguage(System.Reflection.Assembly, System.Resources.UltimateResourceFallbackLocation ByRef)
   at System.Resources.ResourceManager.CommonAssemblyInit()
   at System.Resources.ResourceManager..ctor(System.Type)
   at System.SR.get_ResourceManager()
   at System.SR.InternalGetResourceString(System.String)
   at System.SR.GetResourceString(System.String, System.String)
   at System.Diagnostics.StackTrace.ToString(TraceFormat, System.Text.StringBuilder)
   at System.Diagnostics.StackTrace.ToString(TraceFormat)
   at System.Diagnostics.DebugProvider.Fail(System.String, System.String)
   at System.Globalization.CultureInfo.get_InvariantCulture()
   at System.Resources.ManifestBasedResourceGroveler.GetNeutralResourcesLanguage(System.Reflection.Assembly, System.Resources.UltimateResourceFallbackLocation ByRef)
   at System.Resources.ResourceManager.CommonAssemblyInit()
   at System.Resources.ResourceManager..ctor(System.Type)
   at System.SR.get_ResourceManager()
   at System.SR.InternalGetResourceString(System.String)
   at System.SR.GetResourceString(System.String, System.String)
   at System.DllNotFoundException..ctor()
   at System.Globalization.GlobalizationMode.GetGlobalizationInvariantMode()
   at System.Globalization.GlobalizationMode..cctor()
   at System.Globalization.CultureData.CreateCultureWithInvariantData()
   at System.Globalization.CultureData.get_Invariant()
   at System.Globalization.CultureInfo..cctor()
   at System.Globalization.CultureInfo.get_InvariantCulture()
   at System.Resources.ManifestBasedResourceGroveler.GetNeutralResourcesLanguage(System.Reflection.Assembly, System.Resources.UltimateResourceFallbackLocation ByRef)
   at System.Resources.ResourceManager.CommonAssemblyInit()
   at System.Resources.ResourceManager..ctor(System.Type)
   at System.SR.get_ResourceManager()
   at System.SR.InternalGetResourceString(System.String)
   at System.SR.GetResourceString(System.String, System.String)
   at System.IO.FileNotFoundException.ToString()
artifacts/tests/coreclr/Linux.arm.Checked/JIT/CodeGenBringUpTests/ArrayExc_d/ArrayExc_d.sh: line 275:  1327 Aborted                 $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
Expected: 100
Actual: 134
END EXECUTION - FAILED
```
</details>

Looks like string resources are missing, not sure if that is the same case as CI, but return code is same as seen in [helix log](https://helix.dot.net/api/2019-06-17/jobs/5016ddc9-621c-41c7-a589-34adade1a8d5/workitems/Interop/console): 134. Can we pinpoint to something concrete in the patch with this info? I can switch to debug build, if that would give more insights.

janvorli commented:
Can you please check if the System.Globalization.Native.so is present and valid?

am11 commented:
The nearest System.Globalization.Native.so is present at `artifacts/bin/coreclr/Linux.arm.Checked/crossgen2`. I copied `System.Globalization.Native.so` and `System.Runtime.dll` from crossgen2 to ` artifacts/bin/coreclr/Linux.arm.Checked/` and reran `ArrayExc_d.sh` and it passed.

Also ran all hardwareintrinsic tests successfully, after copying an assembly from different location: ` cp artifacts/bin/runtime/netcoreapp5.0-Linux-Release-arm/System.Runtime.Intrinsics.Experimental.dll  artifacts/bin/coreclr/Linux.arm.Checked/`. One or more of these are failing with exit code 134: https://helix.dot.net/api/2019-06-17/jobs/ba003676-b60a-4de5-987e-c31b7bfcd86b/workitems/JIT.HardwareIntrinsics/console, which is likely "file not found"; then it sounds like this PR is expected to have some missing copy step?

am11 commented:
The Helix runs do not give more information (full console output which we see on device), or maybe I am looking at a wrong place? 🤔

janvorli commented:
@am11 the System.Globalization.Native.so should not be located in the artifacts/bin/coreclr/Linux.arm.Checked. It is not built in the coreclr part of the repo anymore. Now it is built in libraries. 

The reason why it fixed the test for you is that the testing combines files from artifacts/bin/coreclr/Linux.arm.Checked and from the libraries into the artifacts/tests/coreclr/Linux.arm.Checked/Tests/Core_Root directory.
But it should get there together with other stuff built in libraries (e.g. System.Native.so, System.IO.Ports.Native.so, etc). Maybe these could give you a hint.

janvorli commented:
@am11 the copying from libraries happens in this step in runtest.proj:
```xml
  <Target Name="CreateTestOverlay">
    <MSBuild Projects="$(MSBuildProjectFile)"
             Targets="CopyDependencyToCoreRoot"
             Properties="Language=C#;TargetRid=$(TargetRid);RuntimeIdentifier=$(TargetRid)" />
  </Target>
```

It is invoked from build-test.sh using 
```
build_MSBuild_projects "Tests_Overlay_Managed" "${__ProjectDir}/tests/src/runtest.proj" "Creating test overlay" "/t:CreateTestOverlay"
```

I have noticed that your change touched code related to RIDs, I wonder if that could somehow cause the problem.

Anyways, maybe you could look at the logged command line of msbuild.sh invoked from the build_MSBuild_projects for this case and compare it with the one without your changes.

am11 committed:
Unify build_native via eng/native/build-commons
In order to add new platform, architecture, compiler or its newer
version, there are currently multiple places to update, which can be
deduplicated.

This patch unifies:

* directories creation
* prerequisites checking
* `version.c` generation
* native build invocation
* common argument parsing
* building help menu

for various build scripts under coreclr, installer and libraries.

The common entry-point script is now `eng/native/build-commons.sh` and
rest of the scripts under `eng/native` are implementation detail.

Also extracted CMake platform detection in
`eng/native/configureplatform.cmake`, to share with coreclr and
`installer/corehost`.

am11 commented:
/azp run runtime-installer

am11 commented:
Thanks for your tips @janvorli! The issue was with `Libraries Linux arm Release` job not passing `-cross` argument explicitly. In master, we have:

```sh
$RepoRoot/coreclr.sh -cross
$RepoRoot/installer.sh /p:CrossBuild=true
$RepoRoot/libraries.sh    # implictly sets crossbuild=true in build-native.sh
```

pr combines the implementation, favoring the majority and expects explicit arguments, so i have made libraries.sh do the same thing as installer.sh in master (updated ci yaml and build.proj):

```sh
$RepoRoot/coreclr.sh -cross
$RepoRoot/installer.sh /p:CrossBuild=true
$RepoRoot/libraries.sh /p:CrossBuild=true
```

in future, we can make eng/common/build.sh to accept `-cross` as first-class argument for cross compilation. This PR focuses on converging internal implementation.

am11 commented:
WinNT checked failures are unrelated (test timeout). This PR is ready for review. Could you please request `runtime-installer` jobs? It was triggered automatically in last run, but missed in this one. Thanks!

janvorli commented on src/installer/corehost/build.sh:
```diff
@@ -1,282 +1,103 @@
+__CompilerMinorVersion=
+__CrossBuild=0
+__IsMSBuildOnNETCoreSupported=0
+__NumProc=1
```
Why 1?

am11 replied:
`__NumProc=1` was copied from `src/libraries/Native/build-native.sh`. But coreclr/build.sh and coreclr/bulid-native.sh do not explicitly declare it (and let the commons file define it). I will remove it from here.

janvorli commented on src/libraries/Native/build-native.cmd:
```diff
@@ -127,6 +127,11 @@ may help to copy its "DIA SDK" folder into "%VSINSTALLDIR%" manually, then try a
 
 :GenVSSolution
+:: generate version file
+powershell -NoProfile -ExecutionPolicy ByPass -NoLogo -File "%__repoRoot%\eng\common\msbuild.ps1" /clp:nosummary %__ArcadeScriptArgs%^
```
I wonder why was the move from the build-native.proj to this place needed. I am no msbuild expert, so I am just curious.

janvorli replied:
Also, I cannot see similar change in the build-native.sh, so it looks suspicious.

am11 replied:
For Unix, this code is moved to `eng/native/build-commons.sh` in `build_native()` function. I haven't done much work on Windows scripts in this PR, this was just to align it with `coreclr/build.cmd`. In future we can consolidate batch scripts as well.

am11 commented on src/coreclr/build-test.sh:
```diff
@@ -673,42 +592,26 @@ __TestDir="$__ProjectDir/tests"
-
-if [ $__PortableBuild == 0 ]; then
-    __CommonMSBuildArgs="$__CommonMSBuildArgs /p:PortableBuild=false"
+    HOME="$__ProjectDir"/temp_home
```
It is same as line 695-696 in master. When i ran a linter (shellcheck) on it, it suggested to separate export from assignment: [SC2155](https://github.com/koalaman/shellcheck/wiki/SC2155).

janvorli replied:
Ah, I see. I am sorry, I've somehow missed that it was there already.

am11 commented on src/coreclr/build.sh:
```diff
@@ -38,45 +33,23 @@ usage_list+=("-skipnuget: skip NuGet package generation.")
 {
     local OptDataProjectFilePath="$__ProjectRoot/src/.nuget/optdata/optdata.csproj"
-    if [[ ( $__SkipRestoreOptData == 0 ) && ( $__isMSBuildOnNETCoreSupported == 1 ) ]]; then
+    if [ "$__SkipRestoreOptData" = 0 ] && [ "$__IsMSBuildOnNETCoreSupported" = 1 ]; then
```
single `if [ expr ]` is a short form of `if test expr` which is posix shell complaint and uses single `=` for equality check. we are inconsistently using `if [ x = y]`, `if [[ x == y ]]`, `if [ x == y ]` in various places. i just picked up the posix shell flavor and used it everywhere.

janvorli replied:
Could we still use the == instead of = or could it break some shells? I was trying to find a list of comparison operators for posix shells and while I have found some documents on the syntax including grammar, the operators were not there. Do you have a pointer to a document that I can read on that?

am11 replied:
* test(1): https://linux.die.net/man/1/test
* [The Unix Shell's Humble If](https://thoughtbot.com/blog/the-unix-shells-humble-if)

these indicate that `=` is used for string comparison and `-eq` is used for numeric comparison. We can also use `(( x == 1 ))` for numeric comparison.

However, second link also mentions:

```
 Best Practice: If you have no reason to target POSIX shell, use [[
Bracket-Bracket

[[ comes with the following features over the normal test command:

*    Use familiar ==, >=, and <= operators
```

which is the case here: `#!/usr/bin/env bash`.

janvorli replied:
Thank you for the pointers. I would then prefer to unify on the `[[` instead of `[`.

am11 commented on src/coreclr/build.sh:
```diff
@@ -526,12 +394,8 @@ handle_arguments() {
-
         *)
-            __UnprocessedBuildArgs+=("$1")
+            __UnprocessedBuildArgs="$__UnprocessedBuildArgs $1"
```
We are inconsistent in using array:

* src/coreclr/tests/runtesttilstable.sh
* eng/common/cross/build-rootfs.sh
* src/libraries/Native/build-native.sh

are using this concatenation approach, so i aligned it here. Between `+=("$1")` and `="$__UnprocessedBuildArgs $1"` I wold pick latter because it is posix complaint and array is not needed for simple cases like these. in particular, posix shell does not define array by it self.

janvorli replied:
Thank you for the explanation, it makes sense.

am11 commented on src/coreclr/build.sh:
```diff
@@ -472,7 +340,7 @@ handle_arguments() {
         ignorewarnings|-ignorewarnings)
             __IgnoreWarnings=1
-            __cmakeargs="$__cmakeargs -DCLR_CMAKE_WARNINGS_ARE_ERRORS=OFF"
+            __CMakeArgs="-DCLR_CMAKE_WARNINGS_ARE_ERRORS=OFF $__CMakeArgs"
```
CMakeArgs and CMakeExtraArgs are now merged into just one CMakeArgs flat list.
If user specifies, e.g. `build.sh -cmakeargs -DCLR_CMAKE_WARNINGS_ARE_ERRORS=ON`, that will come after all the other defines supported by script. This way we can override those properties from command-line as well which were specified by the script (last define takes the precedence).

janvorli replied:
Ah, got it.

am11 commented on src/coreclr/configurecompiler.cmake:
```diff
@@ -8,194 +8,12 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
-
-# All code we build should be compiled as position independent
-check_pie_supported(OUTPUT_VARIABLE PIE_SUPPORT_OUTPUT LANGUAGES CXX)
```
This code was moved to `eng/native/configureplatform.cmake` which is imported by this file and in installer (both were setting PIE flag). I have verified that libcoreclr.so and corerun are compiled as position independent code.

am11 committed:
Use if [[ cond ]] in all places

am11 committed:
Address CR feedback

am11 committed:
Fix quotes

am11 commented:
@janvorli, I do not have permission to request installer pipeline run, which was running earlier (total 168 checks were executed) but not anymore. Could you please request:
`/azp run runtime-installer`
Also, could you manually re-run `runtime (Test crossgen-comparison Linux arm checked)`, which has been cancelled due to unknown reason? This job was failing earlier due to a misplaced quote, so I wanted to make sure if f37f9b4 has fixed it.

janvorli commented:
/azp run runtime-installer

am11 commented:
Just noticed that runtime-installer pipeline no longer exist: https://dev.azure.com/dnceng/public/_build?pipelineNameFilter=runtime (or maybe its `Service Connection` is dropped due to expired token). Nevertheless, 10 Installer jobs were ran in `runtime` pipeline.

am11 committed:
Merge dotnet/master and resolve conflicts

am11 commented:
@janvorli, there were merge conflict with recent renaming in configurecompiler.cmake file (in the portions which this PR moves to `env/native/configureplatform.cmake` to share between coreclr and installer). Windows NT failures are unrelated `Condition(s) not met: "NotElevatedAndSupportsEventLogs"` and `The "DownloadFile" task failed unexpectedly.`. Could you please give it another pass?

am11 committed:
Merge  dotnet/master and resolve merge conflicts

am11 commented:
There were some more merge conflicts since the previous resolution.

> Linux arm failure is unrelated: `The TLS connection was non-properly terminated.` while fetching `dotnet-install.sh`.

@janvorli, seems like more and more PRs are coming in touching these files. Can this get a turn before they get merged? had to revalidate property values with grep/diff tools. 😅

janvorli commented:
I am sorry for the hassle with resolving the conflicts. Let's get this in. I've added a few cosmetic micro nit comments yesterday, unfortunately I've somehow forgotten push the button to send out the comments. We can fix them later.

janvorli reviewed (approved):
As I've said, please don't care about the new comments now so that we can get it in as soon as possible to avoid further conflicts.

am11 commented on eng/native/build-commons.sh:
```diff
@@ -0,0 +1,479 @@
+
+    if [[ ( "$__HostOS" == "Linux" )  && ( "$__HostArch" == "x64" || "$__HostArch" == "arm" || "$__HostArch" == "arm64" ) ]]; then
+        __IsMSBuildOnNETCoreSupported=1
+    elif [[ "$__HostArch" == "x64" && ( "$__HostOS" == "OSX" || "$__HostOS" == "FreeBSD" ) ]]; then
```
Sure, i will fix these in a followup PR.

am11 replied:
Fixed in https://github.com/dotnet/runtime/pull/2071.
